sigstore/_errors.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2022 The Sigstore Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Exceptions.
+"""
+
+import sys
+from textwrap import dedent
+from typing import Any, Mapping
+
+
+class Error(Exception):
+    """Base sigstore exception type. Defines helpers for diagnostics."""
+
+    def diagnostics(self) -> str:
+        """Returns human-friendly error information."""
+
+        return """An issue occurred."""
+
+    def print_and_exit(self, raise_error: bool = False) -> None:
+        """Prints all relevant error information to stderr and exits."""
+
+        remind_verbose = (
+            "Raising original exception:"
+            if raise_error
+            else "For detailed error information, run sigstore with the `--verbose` flag."
+        )
+
+        print(f"{self.diagnostics()}\n{remind_verbose}", file=sys.stderr)
+
+        if raise_error:
+            # don't want "during handling another exception"
+            self.__suppress_context__ = True
+            raise self
+
+        sys.exit(1)
+
+
+class NetworkError(Error):
+    """Raised when a connectivity-related issue occurs."""
+
+    def diagnostics(self) -> str:
+        return """A network issue occurred.
+
+        Check your internet connection and try again.
+        """
+
+
+class TUFError(Error):
+    """Raised when a TUF error occurs."""
+
+    def __init__(self, message: str):
+        self.message = message
+
+    from tuf.api import exceptions
+
+    _details: Mapping[Any, str] = {
+        exceptions.DownloadError: NetworkError().diagnostics()
+    }
+
+    def diagnostics(self) -> str:
+        details = TUFError._details.get(
+            type(self.__context__),
+            "Please report this issue at <https://github.com/sigstore/sigstore-python/issues/new>.",
+        )
+
+        return f"""{self.message}.
+
+        {details}
+        """
+
+
+class MetadataError(Error):
+    """Raised when TUF metadata does not conform to the expected structure."""
+
+    def diagnostics(self) -> str:
+        return f"""{str(self)}."""
+
+
+class VerificationError(Error):
+    """Raised when the verifier returns a `VerificationFailure` result."""
+
+    # HACK(tnytown): Importing this at the top of the module causes circular import issues.
```

Comment:
This is suspicious, I would appreciate guidance on how we can make it better 😬

Reply:
I just pushed [fc70655](https://github.com/sigstore/sigstore-python/pull/531/commits/fc706557c74e866faf4dc514b5a1b079a7e09300). Feel free to revert it if you'd prefer it some other way.

I fixed the circular import by moving `VerificationError` to `_cli.py`. I think this makes sense because this is a specialised error type that's only used in that module (kinda like how we define `IdentityError` in `oidc.py` since we only use it there).

Reply:
The move looks good, thanks! I wonder if it will keep making sense to declare errors in a module separate from their throw sites. We might run into similar issues with new errors if we ever want to dispatch on the type of a member variable as I did here :(